crates/interpreter/src/exec.rs

```diff
@@ -773,28 +773,47 @@ impl<'m> Thread<'m> {
             Unreachable => return Err(trap()),
             Nop => (),
             Block(b) => self.push_label(self.blocktype(inst, &b), LabelKind::Block),
-            Loop(b) => self.push_label(self.blocktype(inst, &b), LabelKind::Loop(saved)),
+            Loop(b) => self.push_label(
+                self.blocktype(inst, &b),
+                LabelKind::Loop(LoopState {
+                    parser_data: saved,
+                    side_table_idx: *inst.module.index_in_side_table(),
+                }),
+            ),
             If(b) => match self.pop_value().unwrap_i32() {
                 0 => {
-                    self.skip_to_else(inst);
+                    self.jump_from_if(inst);
                     self.push_label(self.blocktype(inst, &b), LabelKind::Block);
                 }
-                _ => self.push_label(self.blocktype(inst, &b), LabelKind::If),
+                _ => {
+                    *inst.module.index_in_side_table() += 1;
+                    self.push_label(self.blocktype(inst, &b), LabelKind::If);
+                }
             },
             Else => {
-                self.skip_to_end(inst, 0);
+                self.jump_to_end(inst, None);
                 return Ok(self.exit_label());
             }
             End => return Ok(self.exit_label()),
-            Br(l) => return Ok(self.pop_label(inst, l)),
+            Br(l) => return Ok(self.pop_label(inst, l, None)),
             BrIf(l) => {
                 if self.pop_value().unwrap_i32() != 0 {
-                    return Ok(self.pop_label(inst, l));
+                    return Ok(self.pop_label(inst, l, None));
                 }
+                *inst.module.index_in_side_table() += 1;
             }
             BrTable(ls, ln) => {
                 let i = self.pop_value().unwrap_i32() as usize;
-                return Ok(self.pop_label(inst, ls.get(i).cloned().unwrap_or(ln)));
+                let ls_idx = i;
+                let mut is_last = false;
+                if i >= ls.len() {
+                    is_last = true;
+                }
+                return Ok(self.pop_label(
+                    inst,
+                    ls.get(i).cloned().unwrap_or(ln),
+                    Option::from((ls_idx, is_last)),
```

Comment:
A bit hacky but saves the 5 lines above. (The compiler should factorize the common sub-expression.)
```suggestion
                    ls.get(i).map(|_| i),
```

Reply:
Does this comment still apply?

Reply:
I see. Thanks!